### ISSUE TITLE
Offline backup not working (permissions)

### ISSUE THREAD

**Issue by mschindhelm** [appscode/stash#534]
Online backups are working for us but if I define offline backups the init container crashes with the following log output:

`E0723 08:15:34.625349       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:35.625569       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:35.630228       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:36.630463       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:36.631793       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:37.631992       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:37.632887       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:38.633104       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:38.634381       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:39.634594       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:39.635961       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:40.636202       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:40.637910       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:41.638134       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:41.639490       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:42.639710       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:42.642119       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:43.642350       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:43.643769       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:44.643993       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:44.645318       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:45.645545       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:45.650127       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:46.650353       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:46.651468       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"
I0723 08:15:47.651727       1 reflector.go:240] Listing and watching *v1alpha1.Restic from github.com/appscode/stash/client/informers/externalversions/factory.go:74
E0723 08:15:47.652942       1 reflector.go:205] github.com/appscode/stash/client/informers/externalversions/factory.go:74: Failed to list *v1alpha1.Restic: restics.stash.appscode.com is forbidden: User "system:serviceaccount:inf:default" cannot list restics.stash.appscode.com in the namespace "inf"`

I deployed stash in the custom stash namespace and defined an offline backup at the inf namespace. Did I miss something to setup regarding RBAC ?

**Comment by hossainemruz** [comment]
@mschindhelm what version of stash are you using?
How did you install stash? Via helm or script?
It seems some RBAC permission are missing.

**Comment by mschindhelm** [comment]
@hossainemruz : Version is:
`# kubectl exec -it $POD_NAME -c operator -n $POD_NAMESPACE stash version
I0723 09:09:29.665209      61 logs.go:19] FLAG: --alsologtostderr="false"
I0723 09:09:29.665567      61 logs.go:19] FLAG: --enable-analytics="true"
I0723 09:09:29.665582      61 logs.go:19] FLAG: --help="false"
I0723 09:09:29.665595      61 logs.go:19] FLAG: --log_backtrace_at=":0"
I0723 09:09:29.665605      61 logs.go:19] FLAG: --log_dir=""
I0723 09:09:29.665609      61 logs.go:19] FLAG: --logtostderr="false"
I0723 09:09:29.665614      61 logs.go:19] FLAG: --short="false"
I0723 09:09:29.665618      61 logs.go:19] FLAG: --stderrthreshold="0"
I0723 09:09:29.665623      61 logs.go:19] FLAG: --v="0"
I0723 09:09:29.665628      61 logs.go:19] FLAG: --vmodule=""
Version = 0.7.0
VersionStrategy = tag
Os = alpine
Arch = amd64
CommitHash = 705ecd09ddde608b9a66ed248acbc8229cfe8c24
GitBranch = HEAD
GitTag = 0.7.0
CommitTimestamp = 2018-05-29T00:35:12
I0723 09:09:29.803162      61 main.go:26] [Exiting Stash Main]`
`kubectl exec -it $POD_NAME -c operator -n $POD_NAMESPACE restic version
restic 0.8.3
compiled with go1.10 on linux/amd64`
I installed it using the following command line:
`curl -fsSL https://raw.githubusercontent.com/appscode/stash/0.7.0/hack/deploy/stash.sh | bash -s -- --run-on-master --namespace=stash`

**Comment by hossainemruz** [comment]
@mschindhelm can you share workload yaml(Deployment/Statefulset) and Restic yaml ?

**Comment by mschindhelm** [comment]
@hossainemruz sure:
workload:
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
  labels:
    app: inventory
  name: inventory
  namespace: inf
  selfLink: /apis/extensions/v1beta1/namespaces/inf/deployments/inventory
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: inventory
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: inventory
    spec:
      containers:
      - env:
        <...>
        image: mysql:5.6
        imagePullPolicy: IfNotPresent
        name: inventory-mysql
        ports:
        - containerPort: 3306
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/lib/mysql
          name: mysql-persistent-storage
      - env:
        <...>
        image: docker.internal.local/docker-inf/snipe-it:4.4.1
        imagePullPolicy: IfNotPresent
        name: snipe-it
        ports:
        - containerPort: 80
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/lib/snipeit
          name: inventory-persistent-storage
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: docker.internal.local
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: mysql-persistent-storage
        persistentVolumeClaim:
          claimName: pvcsc-inventory-mysql
      - name: inventory-persistent-storage
        persistentVolumeClaim:
          claimName: pvcsc-inventory-data
status: { }

Restic:
apiVersion: stash.appscode.com/v1alpha1
kind: Restic
metadata:
  name: inventory
spec:
  selector:
    matchLabels:
      app: inventory
  type: offline
  fileGroups:
  - path: /var/lib/mysql
    retentionPolicyName: 'keep-last-5'
  - path: /var/lib/snipeit
    retentionPolicyName: 'keep-last-5'
  backend:
    local:
      mountPath: /backup
      hostPath:
        path: /mnt/backup
    storageSecretName: restic-inf
  schedule: '@midnight'
  volumeMounts:
  - mountPath: /var/lib/mysql
    name: mysql-persistent-storage
  - mountPath: /var/lib/snipeit
    name: inventory-persistent-storage
  retentionPolicies:
  - name: 'keep-last-5'
    keepLast: 5
    prune: true

**Comment by hossainemruz** [comment]
One more question, the log you have provided is it from stash operator or stash init container?

**Comment by mschindhelm** [comment]
Its from the init container.

**Comment by hossainemruz** [comment]
@mschindhelm thank you for reporting. We could not recreate this RBAC issue. If you can recreate it, please let us know the steps.

However, while debugging this issue we have found another major issue https://github.com/appscode/stash/issues/535 with offline backup. We are actively working to fix this. This fix will be available in the next release.

In the meantime, please make sure that the `Restic` and `Deployment` are in the same namespace. The yaml's you have provided show that `Deployment` is in `inf` namespace where `Restic` is in `default` namespace. All permission given to stash init container is limited within the `Deployment`'s namespace. So, it would not be able to `Get/List` Restic from other namespaces.

**Comment by mschindhelm** [comment]
I currently unsuccessful in recreating the issue. Maybe it only was a temporary issue.
On recreation of the deployment and the restic ressource, the init container finishes with the following result:
```
kubectl -n inf logs inventory-674bf646f7-d99fb stash
I0724 07:31:59.501357       1 logs.go:19] FLAG: --alsologtostderr="false"
I0724 07:31:59.501564       1 logs.go:19] FLAG: --burst="100"
I0724 07:31:59.501580       1 logs.go:19] FLAG: --docker-registry="appscode"
I0724 07:31:59.501586       1 logs.go:19] FLAG: --enable-analytics="true"
I0724 07:31:59.501590       1 logs.go:19] FLAG: --enable-rbac="true"
I0724 07:31:59.501594       1 logs.go:19] FLAG: --help="false"
I0724 07:31:59.501598       1 logs.go:19] FLAG: --image-tag="0.7.0"
I0724 07:31:59.501602       1 logs.go:19] FLAG: --kubeconfig=""
I0724 07:31:59.501608       1 logs.go:19] FLAG: --log_backtrace_at=":0"
I0724 07:31:59.501617       1 logs.go:19] FLAG: --log_dir=""
I0724 07:31:59.501621       1 logs.go:19] FLAG: --logtostderr="false"
I0724 07:31:59.501626       1 logs.go:19] FLAG: --master=""
I0724 07:31:59.501642       1 logs.go:19] FLAG: --pushgateway-url="http://stash-operator.stash.svc:56789"
I0724 07:31:59.501656       1 logs.go:19] FLAG: --qps="100"
I0724 07:31:59.501666       1 logs.go:19] FLAG: --restic-name="inventory"
I0724 07:31:59.501672       1 logs.go:19] FLAG: --resync-period="5m0s"
I0724 07:31:59.501680       1 logs.go:19] FLAG: --run-via-cron="false"
I0724 07:31:59.501684       1 logs.go:19] FLAG: --scratch-dir="/tmp"
I0724 07:31:59.501688       1 logs.go:19] FLAG: --stderrthreshold="0"
I0724 07:31:59.501692       1 logs.go:19] FLAG: --v="3"
I0724 07:31:59.501696       1 logs.go:19] FLAG: --vmodule=""
I0724 07:31:59.501700       1 logs.go:19] FLAG: --workload-kind="Deployment"
I0724 07:31:59.501705       1 logs.go:19] FLAG: --workload-name="inventory"
W0724 07:31:59.638752       1 client_config.go:533] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0724 07:31:59.686129       1 backup.go:91] Skipping backup. Reason: Backup type offline and replica > 1
I0724 07:31:59.686152       1 backup.go:106] [Exiting Stash Backup]
I0724 07:31:59.686168       1 main.go:26] [Exiting Stash Main]
```
why does the stash init container think that the replica count is bigger than one ?

**Comment by hossainemruz** [comment]
Oops!!!. This log line actually misleading.
Stash offline backup works in following steps.
1. Stash operator inject init container to all workload replicas.
2. A `CronJob` run at scheduled time.
3. The `CronJob` scale down the workload to 0 replica and store original replica count in workload annotation.
4. Then it scale up the workload to 1 replica.
5. This replica pod start with init container. Check that there are no other replica pod running. If it is the only replica running then it take backup. Otherwise skip taking backup(data consistency issue).
6. Once backup is complete, this init container read original replica count from annotation and scale up the workload to original replica number.

Here, two situation to be handle
1. When stash operator find a matching `Restic` for a workload, it patch the workload to inject init container. So, the all workload's pod restart with init container. At this time, we have to skip taking backup. Because, all pods has started with init container. We can not guarantee that any of the pods are not in running state while another pod is taking backup.
2. When a pod start with init container for first time, it does not have original replica count in annotation. Because, scheduled `CronJob` hasn't run yet. In this case we have to skip taking backup. Because, it will panic when it will try to scale up the workload after backup is completed.

This two cases are handle in [this](https://github.com/appscode/stash/blob/705ecd09ddde608b9a66ed248acbc8229cfe8c24/pkg/cmds/backup.go#L90) condition but we have logged only for first case.

Thank you for noticing that. We are fixing now. However, we recommended not to use offline backup in production cluster due to issue #535 . We, will release a new version soon with the fix.
